config/translate/translate.go

```diff
@@ -0,0 +1,202 @@
+// Copyright 2019 Red Hat, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package translate
+
+import (
+	"fmt"
+	"reflect"
+)
+
+func isPrimitive(k reflect.Kind) bool {
+	switch k {
+	case reflect.Bool,
+		reflect.Int,
+		reflect.Int8,
+		reflect.Int16,
+		reflect.Int32,
+		reflect.Int64,
+		reflect.Uint,
+		reflect.Uint8,
+		reflect.Uint16,
+		reflect.Uint32,
+		reflect.Uint64,
+		reflect.Uintptr,
+		reflect.Float32,
+		reflect.Float64,
+		reflect.Complex64,
+		reflect.Complex128,
+		reflect.String:
+		return true
+	default:
+		return false
+	}
+}
+
+func isInvalidInConfig(k reflect.Kind) bool {
+	switch k {
+	// These don't serialize well, shouldn't be in the generated structs anyway
+	case reflect.Array,
+		reflect.Chan,
+		reflect.Func,
+		reflect.Interface,
+		reflect.Map,
+		reflect.UnsafePointer,
+		reflect.Invalid:
+		return true
+	default:
+		return false
+	}
+}
+
+func sameType(t1, t2 reflect.Type) bool {
+	k1 := t1.Kind()
+	k2 := t2.Kind()
+	if k1 != k2 {
+		return false
+	}
+	switch {
+	case isPrimitive(k1):
+		return true
+	case isInvalidInConfig(k1):
+		panic("Encountered invalid kind in config. This is a bug, please file a report")
+	case k1 == reflect.Ptr || k1 == reflect.Slice:
+		return sameType(t1.Elem(), t2.Elem())
+	case k1 == reflect.Struct:
+		return sameStruct(t1, t2)
+	default:
+		panic("Encountered unknown kind in config. This is a bug, please file a report")
+	}
+}
+
+// precondition: t1, t2 are both of Kind 'struct'
+func sameStruct(t1, t2 reflect.Type) bool {
+	if t1.NumField() != t2.NumField() || t1.Name() != t2.Name() {
+		return false
+	}
+	for i := 0; i < t1.NumField(); i++ {
+		t1f := t1.Field(i)
+		t2f, ok := t2.FieldByName(t1f.Name)
+
+		if !ok || !sameType(t1f.Type, t2f.Type) {
+			return false
+		}
+	}
+	return true
+}
+
+// checks that t could reasonably be the type of a translator function
+func probablyValidTranslator(t reflect.Type) bool {
```

Comment:
Bikeshed: how about `couldBeValidTranslator`? We want `true` to mean "maybe", but `false` to mean "definitely not", but the "probably" hides that a bit.